src/github.com/matrix-org/dendrite/clientapi/writers/createroom.go

```diff
@@ -117,12 +130,141 @@ func createRoom(req *http.Request, roomID string) util.JSONResponse {
+		}
+		ev, err := buildEvent(&builder, builtEventMap, cfg)
+		if err != nil {
+			return util.ErrorResponse(err)
+		}
+
+		if err := gomatrixserverlib.Allowed(*ev, &authEvents); err != nil {
+			return util.ErrorResponse(err)
+		}
+
+		builtEventMap[common.StateTuple{e.Type, *e.StateKey}] = ev
+		builtEvents = append(builtEvents, ev)
 
-	// f.e event:
-	// - validate required keys/types (EventValidator in synapse)
-	// - set additional keys (displayname/avatar_url for m.room.member)
-	// - set token(?) and txn id
-	// - then https://github.com/matrix-org/synapse/blob/v0.19.2/synapse/handlers/message.py#L419
+	}
+
+	return util.JSONResponse{
+		Code: 200,
+		JSON: builtEvents,
+	}
+}
+
+// buildEvent fills out auth_events for the builder then builds the event
+func buildEvent(builder *gomatrixserverlib.EventBuilder,
+	events map[common.StateTuple]*gomatrixserverlib.Event,
+	cfg config.ClientAPI) (*gomatrixserverlib.Event, error) {
+
+	eventsNeeded, err := gomatrixserverlib.StateNeededForEventBuilder(builder)
+	if err != nil {
+		return nil, err
+	}
+	builder.AuthEvents = authEventsFromStateNeeded(eventsNeeded, events)
+	eventID := fmt.Sprintf("$%s:%s", util.RandomString(16), cfg.ServerName)
+	now := time.Now()
+	event, err := builder.Build(eventID, now, cfg.ServerName, cfg.KeyID, cfg.PrivateKey)
+	if err != nil {
+		return nil, fmt.Errorf("cannot build event %s : Builder failed to build. %s", builder.Type, err)
+	}
+	return &event, nil
+}
+
+func authEventsFromStateNeeded(eventsNeeded gomatrixserverlib.StateNeeded,
+	events map[common.StateTuple]*gomatrixserverlib.Event) (authEvents []gomatrixserverlib.EventReference) {
+
+	// These events are only "needed" if they exist, so if they don't exist we can safely ignore them.
+	if eventsNeeded.Create {
+		ev := events[common.StateTuple{"m.room.create", ""}]
+		if ev != nil {
+			authEvents = append(authEvents, ev.EventReference())
+		}
+	}
+	if eventsNeeded.JoinRules {
+		ev := events[common.StateTuple{"m.room.join_rules", ""}]
+		if ev != nil {
+			authEvents = append(authEvents, ev.EventReference())
+		}
+	}
+	if eventsNeeded.PowerLevels {
+		ev := events[common.StateTuple{"m.room.power_levels", ""}]
+		if ev != nil {
+			authEvents = append(authEvents, ev.EventReference())
+		}
+	}
+
+	for _, userID := range eventsNeeded.Member {
+		ev := events[common.StateTuple{"m.room.member", userID}]
+		if ev != nil {
+			authEvents = append(authEvents, ev.EventReference())
+		}
+	}
+	for _, token := range eventsNeeded.ThirdPartyInvite {
+		ev := events[common.StateTuple{"m.room.member", token}]
+		if ev != nil {
+			authEvents = append(authEvents, ev.EventReference())
+		}
+	}
+	return
+}
+
+type authEventProvider struct {
+	events map[common.StateTuple]*gomatrixserverlib.Event
+}
+
+func (a *authEventProvider) Create() (ev *gomatrixserverlib.Event, err error) {
+	return a.fetch(common.StateTuple{"m.room.create", ""})
```

Comment:
Oh, it did before when I needed to check it for `nil` to return an `error`, but yeah it's obsolete now.